File: problems/p9/ocaml/solve.ml
let iota m n =
  let rec loop x acc = if x < m then loop (x + 1) (x :: acc) else List.rev acc in
  loop n []

let pythagoreans m =
  assert (2 <= m);
  iota m 1
  |> List.map (fun n -> (m * m - n * n, 2 * m * n, m * m + n * n))

let solve summary =
  iota summary 2 |> List.concat_map pythagoreans |> List.find (fun (a, b, c) -> a + b + c = summary)
  
let list_of_tuple = function
  | a, b, c -> [a; b; c]

let () =
  assert (solve 12 |> list_of_tuple |> List.fold_left ( * ) 1 = 60);
  assert (solve 1000 |> list_of_tuple |> List.fold_left ( * ) 1 = 31875000)

File: problems/p9/haskell/solve.hs
import Data.List (find, concatMap)
import Control.Exception (assert)

pythagoreans :: Integer -> [(Integer, Integer, Integer)]
pythagoreans m = [(m * m - n * n, 2 * m * n, m * m + n * n) | n <- [m - 1, m - 2 .. 1]]

solve' :: Integer -> [Integer]
solve' summary = case find (\ (a,b,c) -> a + b + c == summary) $ concatMap pythagoreans [2..] of
  Just (a,b,c) -> [a,b,c]
  Nothing -> []

solve = product . solve'

main :: IO ()
main = do
  assert (solve 12 == 60) $ pure ()
  assert (solve 1000 == 31875000) $ pure ()
